crates/compiler/mono/src/borrow.rs

```diff
@@ -1,119 +1,401 @@
+                (key, signature)
+            })
+            .collect(),
+    };
+
+    // for every proc (by index) a collection of its join points
+    let mut join_points: Vec<_> = std::iter::repeat_with(MutMap::default)
+        .take(procs.len())
+        .collect_in(arena);
+
+    // next we first partition the functions into strongly connected components, then do a
+    // topological sort on these components, finally run the fix-point borrow analysis on each
+    // component (in top-sorted order, from primitives (std-lib) to main)
+
+    let matrix = construct_reference_matrix(arena, procs);
+    let sccs = matrix.strongly_connected_components_all();
+
+    let mut join_point_stack = Vec::new_in(arena);
+    let mut proc_join_points = MutMap::default();
+
+    for (group, _) in sccs.groups() {
+        // This is a fixed-point analysis
+        //
+        // all functions initially own all their parameters
+        // through a series of checks and heuristics, some arguments are set to borrowed
+        // when that doesn't lead to conflicts the change is kept, otherwise it may be reverted
+        //
+        // when the signatures no longer change, the analysis stops and returns the signatures
+
+        loop {
+            let mut modified = false;
+
+            for index in group.iter_ones() {
+                let (_, proc) = procs.iter().nth(index).unwrap();
+                let key = (proc.name.name(), proc.proc_layout(arena));
+
+                if proc.args.is_empty() {
+                    continue;
+                }
+
+                std::mem::swap(&mut proc_join_points, &mut join_points[index]);
+
+                let mut state = State {
+                    args: proc.args,
+                    borrow_signature: *borrow_signatures.procs.get(&key).unwrap(),
+                    join_point_stack,
+                    join_points: proc_join_points,
+                    modified: false,
+                };
+
+                state.inspect_stmt(interner, &mut borrow_signatures, &proc.body);
+
+                // did any proc signature get modified?
+                //
+                // NOTE: this does not directly include updates to join point signatures. The
+                // assumption is that a relevant change in join point signature is immediately
+                // (i.e. no fixpoint is required) reflected in the proc signature.
+                assert_eq!(
+                    state.modified,
+                    borrow_signatures
+                        .procs
+                        .insert(key, state.borrow_signature)
+                        .unwrap()
+                        != state.borrow_signature
+                );
+                modified |= state.modified;
+
+                proc_join_points = state.join_points;
+
+                std::mem::swap(&mut proc_join_points, &mut join_points[index]);
+
+                join_point_stack = state.join_point_stack;
+                join_point_stack.clear();
```

Comment:
Nit: If State takes a reference to `&mut join_points[index]` and has a constructor to do the first swap, I think all of this could be hidden behind a constructor and drop impl. Minorly less error prone, but given state is special just for this code here and probably won't be reused, not particularly important.